### PR TITLE
Feature/36 add cas proxy ticketing

### LINKED ISSUE #36: Allow Proxy-Ticketing from CAS
SonarQube should be empowered to process [CAS Proxy Tickets](https://apereo.github.io/cas/6.0.x/installation/Configuring-Proxy-Authentication.html]. 

One idea is to provide an API endpoint that 
1. receives proxy tickets
2. validates its authenticity
3. issues an authentication token, similar to the existing JWT tokens

Unless the generated token is reused in the same connection it is highly probable that the generated token must be stored and thus be removed after it was invalidated.

This is the partner issue to cloudogu/cockpit#116

### PR DESCRIPTION
**Pull request by ppxl**
Resolve #36 by adding Proxy Ticketing for specially crafted SonarQube requests.

Because of SonarQube peculiarities a proxy ticket request can only successfully granted in the scope of the following requirements:

- CAS must be able to validate proxy ticket under the CAS 3.0 specification
- Sonar-CAS-Plugin must be configured for CAS 3.0 spec as well
- the proxy ticket request must be in the form of a HTTP GET request with a basic authentication header following this form:
   - `Authorization: Basic ${username}:ProxyTicket===:${proxyTicket}` where the credential part is base64 encoded, following the usual [RFC 7617 for basic authentication](https://datatracker.ietf.org/doc/html/rfc7617)

Furthermore, this PR adds crucial information about the proxy ticketing workflow as well as ways to properly enable proxy ticketing via the SonarQube configuration

### CODE REVIEW

File: src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java [hunk 1]
```diff
@@ -84,11 +87,15 @@ String getEmail(Map<String, Object> attributes) {
     }
 
     private String getMailAttribute() {
-        return SonarCasProperties.EMAIL_ATTRIBUTE.getString(config, "mail");
+        String mailAttr = SonarCasProperties.EMAIL_ATTRIBUTE.getString(config, "mail");
+        LOG.debug("Using Sonar CAS Property email attribute: {}", mailAttr);
```
Comment (ppxl) [inline src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java]: This log output should be reverted...

File: src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java [hunk 2]
```diff
@@ -84,11 +87,15 @@ String getEmail(Map<String, Object> attributes) {
...
+        return mailAttr;
     }
 
     String getDisplayName(Map<String, Object> attributes) {
-        return getStringAttribute(attributes, getFullNameAttribute());
+        String fullNameAttr = getStringAttribute(attributes, getFullNameAttribute());
+        LOG.debug("Using Sonar CAS Property display name attribute: {}", fullNameAttr);
```
Comment (ppxl) [inline src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java]: This log output should be reverted...

File: docs/architecture_en.md [hunk 3]
```diff
@@ -116,9 +116,62 @@ At a certain point both are similar because in the end the back-channel logout m
...
+#### Authentication with Proxy Tickets
+
+Proxy tickets are a mechanism for indirect authentication of users, but without using their original password.
+
+In addition to SonarQube, there may be other services that CAS authenticates to. The CAS specification speaks here of *
+services*. A login via proxy ticket is then no longer done via SonarQube in interaction with CAS, but via another
+service: A proxy service. The person using the ticket must instead authenticate himself in interaction with the proxy
+service. This proxy service is then issued a proxy granting ticket, which in turn can be used to request proxy tickets.
+
+Proxy tickets are very similar to CAS service tickets and have a short validity period. The proxy service makes the
```
Comment (ppxl) [inline docs/architecture_en.md:128]: Here the responsible Java class should be listed in order to get a feeling which class is responsible for what

File: src/test/java/org/sonar/plugins/cas/LoginHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+        LoginHandler sut = new LoginHandler(null, null, sessionStoreFactory, null);
+
+        // when
+        String actual = sut.getTicketParameter(request);
+
+        // then
+        assertThat(actual).isEqualTo("ST-012345678");
+        verify(request, times(1)).getParameter("ticket");
+    }
+}
```
Comment (ppxl) [inline src/test/java/org/sonar/plugins/cas/LoginHandlerTest.java]: Add newline here

File: src/test/java/org/sonar/plugins/cas/util/RequestStringerTest.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        String actual = RequestStringer.string(request);
+
+        assertThat(actual).isEqualTo("Request data for URL https://server.com/endpointMethod:\tnull\n" +
+                "Headers:\n" +
+                "no headers found\n" +
+                "Cookies:\n" +
+                "no cookies found\n" +
+                "Parameters:\n");
+    }
+}
```
Comment (ppxl) [inline src/test/java/org/sonar/plugins/cas/util/RequestStringerTest.java]: add newline here

Comment (ppxl) [reply src/test/java/org/sonar/plugins/cas/util/RequestStringerTest.java]: done

File: src/test/java/org/sonar/plugins/cas/LoginHandlerTest.java (hunk 4 above)
Comment (ppxl) [reply src/test/java/org/sonar/plugins/cas/LoginHandlerTest.java]: done

File: docs/architecture_en.md (hunk 3 above)
Comment (ppxl) [reply docs/architecture_en.md:128]: done, also in the german doc translation

File: src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java (hunk 1 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java]: done

File: src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java (hunk 2 above)
Comment (ppxl) [reply src/main/java/org/sonar/plugins/cas/CasAttributeSettings.java]: done

**Review by cbeyer42 — APPROVED** [review]

**Pull request merged**